### PR TITLE
Store replicas in the segments

### LINKED ISSUE #947: Change the segment/node locking system
Change the lock system to enable/bring:
- faster repairs (less contention)
- simpler code
- concurrent repairs on different keyspaces

### PR DESCRIPTION
**Pull request by adejanovski**
This will allow to get more visibility on which nodes are involved in segments, reduce the number of JMX calls to perform when starting a segment.
It is also a first step to build a new segment scheduler that will not rely on JMX metrics for limiting the number of repair per node but rather use backend storage with locks.

### CODE REVIEW

File: src/server/src/main/java/io/cassandrareaper/core/RepairSegment.java [hunk 1]
```diff
@@ -108,6 +111,11 @@ public DateTime getEndTime() {
     return endTime;
   }
 
+  @Nullable
```
Comment (michaelsembwever) [inline src/server/src/main/java/io/cassandrareaper/core/RepairSegment.java]: what does the null state represent, versus an empty collection?

File: src/server/src/main/java/io/cassandrareaper/service/RepairRunService.java [hunk 2]
```diff
@@ -185,7 +196,32 @@ public RepairRun registerRepairRun(
...
+                 .withBaseRange(segment.getBaseRange())
+                 .withTokenRanges(segment.getTokenRanges())
+                 .withReplicas(getDCsByNodeForRepairSegment(targetCluster, segment, repairUnit.getKeyspaceName()))
+                 .build());
+    }
+
+    return segmentsWithReplicas;
+  }
+
+  private Map<String, String> getDCsByNodeForRepairSegment(
```
Comment (michaelsembwever) [inline src/server/src/main/java/io/cassandrareaper/service/RepairRunService.java:214]: my opinion is that this method should be in the `SegmentRunner` class.

and, the existing `getDCsByNodeForRepairSegment` there should be removed as no one is using it now.
https://github.com/thelastpickle/cassandra-reaper/blob/99ff8fc1743106cc3905a916ee44641d0a7ea6b0/src/server/src/main/java/io/cassandrareaper/service/SegmentRunner.java#L750

File: src/server/src/main/java/io/cassandrareaper/service/RepairRunService.java [hunk 3]
```diff
@@ -185,7 +196,32 @@ public RepairRun registerRepairRun(
...
+      Cluster cluster,
+      Segment segment,
+      String keyspace) throws ReaperException {
+
+    JmxProxy jmxConnection = clusterFacade.connect(cluster);
+    // when hosts are coming up or going down, this method can throw an UndeclaredThrowableException
+    Collection<String> nodes = clusterFacade.tokenRangeToEndpoint(cluster, keyspace, segment);
```
Comment (michaelsembwever) [inline src/server/src/main/java/io/cassandrareaper/service/RepairRunService.java]: How do we deal with failure here? Previously we'd just move on to another segment? Now the whole repair is jammed/failed because initialisation fails?

File: src/server/src/main/java/io/cassandrareaper/service/SegmentRunner.java [hunk 4]
```diff
@@ -503,7 +503,7 @@ boolean canRepair(
 
     if (RepairSegment.State.NOT_STARTED == segment.getState()) {
       try {
-        Map<String, String> dcByNode = getDCsByNodeForRepairSegment(coordinator, cluster, segment, keyspace);
+        Map<String, String> dcByNode = segment.getReplicas();
```
Comment (michaelsembwever) [inline src/server/src/main/java/io/cassandrareaper/service/SegmentRunner.java:506]: nice that this is reducing the complexity (time and network calls) involved when determining whether to start a segment!

File: src/server/src/main/java/io/cassandrareaper/service/RepairRunService.java (hunk 3 above)
Comment (adejanovski) [reply src/server/src/main/java/io/cassandrareaper/service/RepairRunService.java]: It would fail indeed.
I can add a retry mechanism to at least give it a few attempts before failing the repair init.
wdyt?

File: src/server/src/main/java/io/cassandrareaper/service/RepairRunService.java (hunk 2 above)
Comment (adejanovski) [reply src/server/src/main/java/io/cassandrareaper/service/RepairRunService.java:214]: I agree that the existing one in the `SegmentRunner` class should be removed, but I don't see a reason why the one from `RepairRunService` should move to `SegmentRunner`. It's used in `RepairRunService.generateSegments()` only and we don't have a SegmentRunner object instanciated at that stage.

File: src/server/src/main/java/io/cassandrareaper/service/RepairRunService.java (hunk 3 above)
Comment (adejanovski) [reply src/server/src/main/java/io/cassandrareaper/service/RepairRunService.java]: Added a retry mechanism

**Comment by adejanovski** [comment]
@michaelsembwever, I've addressed your comments and removed draft state.

File: src/server/src/main/java/io/cassandrareaper/service/RepairRunService.java (hunk 2 above)
Comment (michaelsembwever) [reply src/server/src/main/java/io/cassandrareaper/service/RepairRunService.java:214]: > I don't see a reason why the one from RepairRunService should move to SegmentRunner. It's used in RepairRunService.generateSegments() only and we don't have a SegmentRunner object instanciated at that stage.

Agreed.

File: src/server/src/main/java/io/cassandrareaper/core/RepairSegment.java [hunk 5]
```diff
@@ -53,6 +55,7 @@ private RepairSegment(Builder builder, @Nullable UUID id) {
     this.coordinatorHost = builder.coordinatorHost;
     this.startTime = builder.startTime;
     this.endTime = builder.endTime;
+    this.replicas = builder.replicas;
```
Comment (michaelsembwever) [inline src/server/src/main/java/io/cassandrareaper/core/RepairSegment.java]:
```suggestion
    this.replicas = ImmutableMap.copyOf(builder.replicas);
```

File: src/server/src/main/java/io/cassandrareaper/core/Segment.java [hunk 6]
```diff
@@ -36,13 +37,15 @@
...
   List<RingRange> tokenRanges;
+  Map<String, String> replicas;
 
   private Segment(Builder builder) {
     this.tokenRanges = builder.tokenRanges;
     this.baseRange = builder.tokenRanges.get(0);
     if (builder.baseRange != null) {
       this.baseRange = builder.baseRange;
     }
+    this.replicas = builder.replicas;
```
Comment (michaelsembwever) [inline src/server/src/main/java/io/cassandrareaper/core/Segment.java]:
```suggestion
    this.replicas = ImmutableMap.copyOf(builder.replicas);
```

File: src/server/src/main/java/io/cassandrareaper/service/RepairRunService.java [hunk 7]
```diff
@@ -292,6 +341,7 @@ static int computeGlobalSegmentCount(
                                 .withTokenRanges(Arrays.asList(range.getValue()))
                                 .build(),
                             repairUnit.getId())
+                        .withReplicas(Maps.newHashMap())
```
Comment (michaelsembwever) [inline src/server/src/main/java/io/cassandrareaper/service/RepairRunService.java]:
```suggestion
                        .withReplicas(Collections.EMPTY_MAP)
```
?

File: src/server/src/main/java/io/cassandrareaper/service/RepairRunService.java [hunk 8]
```diff
@@ -185,7 +196,45 @@ public RepairRun registerRepairRun(
...
+
+    return segmentsWithReplicas;
+  }
+
+  private Map<String, String> getDCsByNodeForRepairSegment(
+      Cluster cluster,
+      Segment segment,
+      String keyspace) throws ReaperException {
+
+    final int maxAttempts = 5;
```
Comment (michaelsembwever) [inline src/server/src/main/java/io/cassandrareaper/service/RepairRunService.java]: why `5`? In what time-frame can the 5 subsequent calls fail within?

File: src/server/src/main/java/io/cassandrareaper/service/RepairRunService.java (hunk 3 above)
Comment (michaelsembwever) [reply src/server/src/main/java/io/cassandrareaper/service/RepairRunService.java]: Is this retry mechanism at repair-run initialisation enough? 

What happens when there are range movements during an ongoing repair-run? How do we detect that? How does it fail? Why (should) it fail? Do the docs need to be updated to describe these types of situations?

File: src/server/src/main/java/io/cassandrareaper/service/RepairRunService.java (hunk 8 above)
Comment (adejanovski) [reply src/server/src/main/java/io/cassandrareaper/service/RepairRunService.java]: It's totally arbitrary. What's your take on this?

File: src/server/src/main/java/io/cassandrareaper/service/RepairRunService.java (hunk 3 above)
Comment (adejanovski) [reply src/server/src/main/java/io/cassandrareaper/service/RepairRunService.java]: My take on this is that repair should fail if there's a range movement. Such operations will probably break the guarantees that repair brings.

Comment (adejanovski) [reply src/server/src/main/java/io/cassandrareaper/service/RepairRunService.java]: I made it so that repair fails in case of topology change with an appropriate error message.
Anyway, adding/removing nodes would probably lead to the same result as token ranges that were computed for the segments aren't valid anymore.

wdyt @michaelsembwever ?

File: src/server/src/main/java/io/cassandrareaper/service/RepairRunService.java (hunk 8 above)
Comment (michaelsembwever) [reply src/server/src/main/java/io/cassandrareaper/service/RepairRunService.java]: IMHO it shouldn't be a hardcoded (nor magic) constant, and it should have a sleep in the loop.

Looking at the code I would think that when a node is down it is going to fail 5 times in a row reasonably quickly.

How long can/should this loop run for? How long should we be waiting for a node to become available again? What is this relative to, e.g. other settings in reaper?

File: src/server/src/main/java/io/cassandrareaper/service/RepairRunner.java [hunk 9]
```diff
@@ -465,33 +465,65 @@ private boolean repairSegment(final int rangeIndex, final UUID segmentId, Segmen
...
+      } else if (
+          !Sets.difference(
+            segmentReplicas,
+            potentialCoordinators.stream().collect(Collectors.toSet())).isEmpty()) {
+        LOG.warn(
+            "Segment #{} is faulty, replica set changed since repair was started: [{}, {}]",
+            segmentId,
+            segment.getStartToken(), segment.getEndToken());
+        // This segment has a faulty token range. Abort the entire repair run.
+        synchronized (this) {
```
Comment (michaelsembwever) [inline src/server/src/main/java/io/cassandrareaper/service/RepairRunner.java:512]: why `synchronized` ?

File: src/server/src/main/java/io/cassandrareaper/service/RepairRunner.java [hunk 10]
```diff
@@ -465,33 +465,65 @@ private boolean repairSegment(final int rangeIndex, final UUID segmentId, Segmen
...
+          context.storage.updateRepairRun(
+              context.storage.getRepairRun(repairRunId).get()
+                  .with()
+                  .runState(RepairRun.RunState.ERROR)
+                  .lastEvent(String.format("Replica set changed for segment %s on range [%s, %s]",
+                      segmentId,
+                      segment.getStartToken(),
+                      segment.getEndToken()))
+                  .endTime(DateTime.now())
+                  .build(repairRunId));
```
Comment (michaelsembwever) [inline src/server/src/main/java/io/cassandrareaper/service/RepairRunner.java:522]: This kills/aborts the whole repair run correct?  

What about other segments in it currently running? 

What about other repair runs registered but not yet started?

**Comment by michaelsembwever** [comment]
Couple of questions posted.

Plus,   tests are needed for handling failures around token range movements, as that is a big part of the code change.

File: src/server/src/main/java/io/cassandrareaper/service/RepairRunService.java (hunk 3 above)
Comment (michaelsembwever) [reply src/server/src/main/java/io/cassandrareaper/service/RepairRunService.java]: 👍 

I agree that the whole repair run should fail (quickly and cleanly) on any token range movement.

File: src/server/src/main/java/io/cassandrareaper/service/RepairRunner.java (hunk 9 above)
Comment (adejanovski) [reply src/server/src/main/java/io/cassandrareaper/service/RepairRunner.java:512]: Repair run state updates are always protected by synchronized blocks. That's especially useful to deal with JMX notifications (if we decided to fail the repair run, we don't want a JMX notification to short circuit that in any way).

File: src/server/src/main/java/io/cassandrareaper/service/RepairRunner.java (hunk 10 above)
Comment (adejanovski) [reply src/server/src/main/java/io/cassandrareaper/service/RepairRunner.java:522]: yes, the whole repair will fail.
Running segments won't be impacted by this.

Other repair runs registered but not started will fail as well as soon as they hit a segment whose replica set is no longer valid.

File: src/server/src/main/java/io/cassandrareaper/service/RepairRunService.java (hunk 8 above)
Comment (adejanovski) [reply src/server/src/main/java/io/cassandrareaper/service/RepairRunService.java]: The idea here is rather to handle transient networking issues rather than handling nodes going down.
The way we implement this is that we'll go through `ClusterFacade.connect()` which will pick the fastest node answering in the cluster and then invoke `getRangeToEndpointMap()` on that node.
If we manage connecting to the node, and then it goes down before `getRangeToEndpointMap()`, then the loop will trigger the operation again, picking another node and trying to get the info from it. We won't end up trying to get the information from the same node throughout the retries.
That should be fairly resilient already I think and failing to create the repair run is ok as well (it'll get rescheduled by Reaper or the user will get some feedback and click the button again), so it's probably not worth waiting for too long there.

I can move the max number of attempts to a constant at the top, but I don't think we should make it configurable (probably no need to complexify the config file for this).
Looking at my above description of how things go, I think we could just perform a single retry to cover edge cases and fail fast instead of retrying too much.
wdyt?

**Comment by adejanovski** [comment]
@michaelsembwever, 

I've pushed a few commits adding a unit test for topology changes and reducing the attempts on getting the list of replicas to a single one when generating segments.

Should be good unless you want to further discuss the retry stuff.

**Review by michaelsembwever — APPROVED** [review]
LGTM. just needs the commits squashed and commit msg written up…
